File: Mage/IDPCoordinator.m
//
//  OAuthCoordinator.m
//  MAGE
//
//  Created by William Newman on 5/18/20.
//  Copyright © 2020 National Geospatial Intelligence Agency. All rights reserved.
//

#import "IDPCoordinator.h"
#import "DeviceUUID.h"

@interface IDPCoordinator ()
@property (weak, nonatomic) NSString *url;
@property (assign, nonatomic) NSDictionary *strategy;
@property (weak, nonatomic) UIViewController *viewController;
@property (strong, nonatomic) SFSafariViewController *safariViewController;
@property (strong, nonatomic) id<LoginDelegate> delegate;
@end

@implementation IDPCoordinator

- (instancetype) initWithViewController: (UIViewController *) viewController url: (NSString *) url strategy: (NSDictionary *) strategy delegate: (id<LoginDelegate>) delegate {
    if (self = [super init]) {
        self.url = url;
        self.strategy = strategy;
        self.viewController = viewController;
        self.delegate = delegate;
    }
    
    return self;
}

- (void) start {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(mageAppLinkNotification:) name:@"MageAppLink" object:nil];
    
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"%@?state=mobile", self.url]];
    self.safariViewController = [[SFSafariViewController alloc] initWithURL:url];
    self.safariViewController.modalPresentationStyle = UIModalPresentationPageSheet;
    self.safariViewController.delegate = self;
    self.safariViewController.presentationController.delegate = self;
    [self.viewController presentViewController:self.safariViewController animated:YES completion:nil];
}

- (void)safariViewControllerDidFinish:(SFSafariViewController *)controller {
    [self cleanup];
}

- (void) presentationControllerDidDismiss:(UIPresentationController *)presentationController {
    [self cleanup];
}

- (void) cleanup {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:@"MageAppLink" object:nil];
}

- (void) mageAppLinkNotification:(NSNotification *) notification {
    [self cleanup];

    NSURL *url = [notification object];
    
    if ([url.path containsString:@"authentication"]) {
        NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:false];
        NSURLQueryItem *item = [[components.queryItems filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"name=%@", @"token"]] firstObject];
        [self completeSignin:[item value]];
    } else {
        [self completeSignup];
    }
}

- (void) completeSignup {
    UIAlertController * alert = [UIAlertController
                                 alertControllerWithTitle:@"Account Creation Success"
                                 message:@"Your account has been successfully created.  An administrator must approve your account before you can login"
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
    [self.safariViewController dismissViewControllerAnimated:YES completion:^{
        [self.viewController presentViewController:alert animated:YES completion:nil];
    }];
}

- (void) completeSignin:(NSString *) token {
    NSDictionary *infoDict = [[NSBundle mainBundle] infoDictionary];
    NSString *appVersion = [infoDict objectForKey: @"CFBundleShortVersionString"];
    NSString *buildNumber = [infoDict objectForKey:@"CFBundleVersion"];
    NSDictionary* parameters = @{
                                 @"strategy": self.strategy,
                                 @"token": token,
                                 @"uid": [DeviceUUID retrieveDeviceUUID].UUIDString,
                                 @"appVersion": [NSString stringWithFormat:@"%@-%@", appVersion, buildNumber]
                                 };
    
    [self.safariViewController dismissViewControllerAnimated:YES completion:nil];
    [self.delegate loginWithParameters:parameters withAuthenticationStrategy:[self.strategy objectForKey:@"identifier"] complete:^(AuthenticationStatus authenticationStatus, NSString *errorString) {
        NSLog(@"Authentication complete %ld", (long)authenticationStatus);
    }];
}

@end

File: Mage/Views/IDPCoordinator.swift
//
//  IDPCoordinator.swift
//  MAGE
//
//  Created by Brent Michalski on 8/5/25.
//  Copyright © 2025 National Geospatial Intelligence Agency. All rights reserved.
//

import Foundation
import SafariServices
import UIKit

final class IDPCoordinator: NSObject, SFSafariViewControllerDelegate, UIAdaptivePresentationControllerDelegate {
    private weak var presenter: UIViewController?
    private let url: String
    private let strategy: [String: Any]
    private weak var delegate: IDPCoordinatorDelegate?
    
    private var safariVC: SFSafariViewController?
    private var notificationObserver: NSObjectProtocol?
    
    init(presenter: UIViewController, url: String, strategy: [String: Any], delegate: IDPCoordinatorDelegate?) {
        self.presenter = presenter
        self.url = url
        self.strategy = strategy
        self.delegate = delegate
        super.init()
    }
    
    deinit {
        cleanup()
    }

    func start() {
        notificationObserver = NotificationCenter.default.addObserver(
            forName: Notification.Name("MageAppLink"),
            object: nil,
            queue: .main
        ) { [weak self] notification in
            self?.handleAppLink(notification: notification)
        }
        
        let urlWithState = "\(url)?state=mobile"
        guard let safariURL = URL(string: urlWithState) else { return }
        
        let safariVC = SFSafariViewController(url: safariURL)
        safariVC.modalPresentationStyle = .pageSheet
        safariVC.delegate = self
        safariVC.presentationController?.delegate = self
        self.safariVC = safariVC
        
        presenter?.present(safariVC, animated: true)
    }
    
    func safariViewControllerDidFinish(_ controller: SFSafariViewController) {
        cleanup()
    }
    
    func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
        cleanup()
    }
    
    private func cleanup() {
        if let observer = notificationObserver {
            NotificationCenter.default.removeObserver(observer)
            notificationObserver = nil
        }
    }
    
    private func handleAppLink(notification: Notification) {
        cleanup()
        
        guard let url = notification.object as? URL else { return }
        if url.path.contains("authentication") {
            // Get the token
            guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
                  let tokenItem = components.queryItems?.first(where: { $0.name == "token" }),
                  let token = tokenItem.value else { return }
            
            completeSignIn(token: token)
        } else {
            completeSignUp()
        }
    }
        
    private func completeSignIn(token: String) {
        let infoDict = Bundle.main.infoDictionary
        let appVersion = infoDict?["CFBundleShortVersionString"] as? String ?? ""
        let buildNumber = infoDict?["CFBundleVersion"] as? String ?? ""
        let parameters: [String: Any] = [
            "strategy": strategy,
            "token": token,
            "uid": DeviceUUID.retrieveDeviceUUID()?.uuidString ?? "",
            "appVersion": "\(appVersion)-\(buildNumber)"
        ]
        
        safariVC?.dismiss(animated: true) { [weak self] in
            self?.delegate?.idpCoordinatorDidCompleteSignIn(parameters: parameters)
        }
    }
    
    private func completeSignUp() {
        safariVC?.dismiss(animated: true) { [weak self] in
            self?.delegate?.idpCoordinatorDidCompleteSignUp()
        }
    }
}
